common/network/queues/concurrent_queue.h.inc

```diff
@@ -0,0 +1,95 @@
+#include <pthread.h>
+#include <stdlib.h>
+
+#define DECLARE_STRUCT_CONCURRENT_QUEUE_NODE_OF(TYPE)  \
+  typedef struct concurrent_queue_node_of_##TYPE##_s { \
+    TYPE data;                                         \
+    struct concurrent_queue_node_of_##TYPE##_s *next;  \
+  } concurrent_queue_node_of_##TYPE##_t;
+
+#define DECLARE_VTABLE_CONCURRENT_QUEUE_OF(TYPE)                           \
+  struct concurrent_queue_of_##TYPE##_s;                                   \
+  typedef struct vtable_concurrent_queue_of_##TYPE##_s {                   \
+    TYPE *(*front)(struct concurrent_queue_of_##TYPE##_s * this);          \
+    TYPE *(*back)(struct concurrent_queue_of_##TYPE##_s * this);           \
+    int (*empty)(struct concurrent_queue_of_##TYPE##_s * this);            \
+    size_t (*size)(struct concurrent_queue_of_##TYPE##_s * this);          \
+    void (*push)(struct concurrent_queue_of_##TYPE##_s * this, TYPE data); \
+    void (*pop)(struct concurrent_queue_of_##TYPE##_s * this);             \
+  } vtable_concurrent_queue_of_##TYPE##_t;                                 \
+  extern vtable_concurrent_queue_of_##TYPE##_t const                       \
+      vtable_concurrent_queue_of_##TYPE##_g;
+
+#define DECLARE_STRUCT_CONCURRENT_QUEUE_OF(TYPE)         \
+  typedef struct concurrent_queue_of_##TYPE##_s {        \
+    concurrent_queue_node_of_##TYPE##_t *list;           \
+    concurrent_queue_node_of_##TYPE##_t *front;          \
+    concurrent_queue_node_of_##TYPE##_t *back;           \
+    vtable_concurrent_queue_of_##TYPE##_t const *vtable; \
+    size_t size;                                         \
+    pthread_rwlock_t lock;                               \
```

Comment:
pthread is not available on embedded necessarily.
let's find a sane way to abstract this?

Reply:
Done